tsdb/engine/tsm1/cache.go

```diff
@@ -162,29 +100,69 @@ func (c *Cache) WriteMulti(values map[string][]Value, checkpoint uint64) error {
 	}
 
 	// Enough room in the cache?
-	if c.maxSize > 0 && c.size+uint64(totalSz) > c.maxSize {
+	newSize := c.size + uint64(totalSz)
+	if c.maxSize > 0 && newSize+c.flushingCachesSize > c.maxSize {
 		return ErrCacheMemoryExceeded
 	}
 
 	for k, v := range values {
-		if err := c.write(k, v, checkpoint); err != nil {
-			return err
-		}
+		c.write(k, v)
 	}
+	c.size = newSize
+
 	return nil
 }
 
-// SetCheckpoint informs the cache that updates received up to and including checkpoint
-// can be safely evicted.
-func (c *Cache) SetCheckpoint(checkpoint uint64) error {
+// Snapshot will take a snapshot of the current cache, add it to the slice of caches that
+// are being flushed, and reset the current cache with new values
+func (c *Cache) Snapshot() *Cache {
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	if checkpoint < c.checkpoint {
-		return ErrCacheInvalidCheckpoint
+
+	snapshot := NewCache(c.maxSize)
+	snapshot.store = c.store
+	snapshot.size = c.size
+
+	c.store = make(map[string]*entry)
+	c.size = 0
+
+	c.flushingCaches = append(c.flushingCaches, snapshot)
+	c.flushingCachesSize += snapshot.size
+
+	// sort the snapshot before returning it. The compactor and any queries
+	// coming in while it writes will need the values sorted
+	for _, e := range snapshot.store {
+		if e.needSort {
+			e.values = e.values.Deduplicate()
+			e.needSort = false
+		}
+	}
+
+	return snapshot
+}
+
+// ClearSnapshot will remove the snapshot cache from the list of flushing caches and
+// adjust the size
+func (c *Cache) ClearSnapshot(snapshot *Cache) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	caches := make([]*Cache, 0)
+	cleared := false
+	for _, cache := range c.flushingCaches {
```

Comment:
Instead of making a new slice, couldn't you just find the index of the cache being cleared, and delete it from the existing slice? Then you wouldn't need the `cleared` variable either.